shell/platform/embedder/tests/embedder_a11y_unittests.cc

```diff
@@ -50,6 +97,214 @@ TEST_F(EmbedderA11yTest, A11yTreeIsConsistent) {
+  signal_native_latch.Wait();
+  fml::MessageLoop::GetCurrent().RunExpiredTasksNow();
+  semantics_update_latch.Wait();
+
+  // Dispatch a tap to semantics node 42. Wait for NotifySemanticsAction.
+  fml::AutoResetWaitableEvent notify_semantics_action_latch;
+  notify_semantics_action_callback = [&](Dart_NativeArguments args) {
+    int64_t node_id = 0;
+    Dart_GetNativeIntegerArgument(args, 0, &node_id);
+    ASSERT_EQ(42, node_id);
+
+    int64_t action_id;
+    auto handle = Dart_GetNativeIntegerArgument(args, 1, &action_id);
+    ASSERT_FALSE(Dart_IsError(handle));
+    ASSERT_EQ(static_cast<int32_t>(flutter::SemanticsAction::kTap), action_id);
+
+    Dart_Handle semantic_args = Dart_GetNativeArgument(args, 2);
+    int64_t data;
+    Dart_Handle dart_int = Dart_ListGetAt(semantic_args, 0);
+    Dart_IntegerToInt64(dart_int, &data);
+    ASSERT_EQ(2, data);
+
+    dart_int = Dart_ListGetAt(semantic_args, 1);
+    Dart_IntegerToInt64(dart_int, &data);
+    ASSERT_EQ(1, data);
+    notify_semantics_action_latch.Signal();
+  };
+  std::vector<uint8_t> bytes({2, 1});
+  result = FlutterEngineDispatchSemanticsAction(
+      engine.get(), 42, kFlutterSemanticsActionTap, &bytes[0], bytes.size());
+  ASSERT_EQ(result, FlutterEngineResult::kSuccess);
+  notify_semantics_action_latch.Wait();
+
+  // Disable semantics. Wait for NotifySemanticsEnabled(false).
+  fml::AutoResetWaitableEvent notify_semantics_enabled_latch_3;
+  notify_semantics_enabled_callback = [&](Dart_NativeArguments args) {
+    bool enabled = true;
+    Dart_GetNativeBooleanArgument(args, 0, &enabled);
+    ASSERT_FALSE(enabled);
+    notify_semantics_enabled_latch_3.Signal();
+  };
+  result = FlutterEngineUpdateSemanticsEnabled(engine.get(), false);
+  ASSERT_EQ(result, FlutterEngineResult::kSuccess);
+  notify_semantics_enabled_latch_3.Wait();
+}
+
+TEST_F(EmbedderA11yTest, LegacyEmbedderApiStructsUnchanged) {
+  // New members must not be added to FlutterSemanticsNode
+  // or FlutterSemanticsCustomAction as it would break the ABI of
+  // FlutterSemanticsUpdate.
+  // See: https://github.com/flutter/flutter/issues/121176
+  size_t expected_update_size = 3 * sizeof(size_t) +
+                                sizeof(FlutterSemanticsNode*) +
+                                sizeof(FlutterSemanticsCustomAction*);
+
+  size_t expected_node_size =
+      8 + // Struct padding assuming 64 bit processor
+      3 * sizeof(size_t) + 5 * sizeof(double) + 5 * sizeof(int32_t) +
+      3 * sizeof(const int32_t*) + 6 * sizeof(const char*) +
+      sizeof(FlutterRect) + sizeof(FlutterTextDirection) +
+      sizeof(FlutterSemanticsFlag) + sizeof(FlutterTransformation) +
+      sizeof(FlutterSemanticsAction) + sizeof(FlutterPlatformViewIdentifier);
+
+  size_t expected_action_size = sizeof(size_t) + sizeof(int32_t) +
+                                2 * sizeof(const char*) +
+                                sizeof(FlutterSemanticsAction);
```

Comment:
Is there a better way to "lock" these structs to ensure no new members are added? In particular, `FlutterSemanticsNode` is problematic due to architecture-specific struct padding.

Reply:
https://en.cppreference.com/w/cpp/types/offsetof in a googletest maybe?

Reply:
A quick and dirty (though a bit brittle) hacky approach could be to just grep for the current struct as-is. You could keep a canonical set of text files each with a locked struct, and have a grep check. I'm not proud of suggesting it, but it's not particularly more awful than some of the horrors in the licence script.